storage-proofs/src/zigzag_graph.rs

```diff
@@ -542,4 +542,26 @@ mod tests {
         let parents_cache_lock = zigzag_graph.parents_cache.read().unwrap();
         (*parents_cache_lock)[zigzag_graph.get_cache_index()].len()
     }
+
+    #[test]
+    fn test_shuffle() {
+        let n = 2_u64.pow(10);
+        let d = DEFAULT_EXPANSION_DEGREE as u64;
+        // Use a relatively small value of `n` as Feistel is expensive (but big
+        // enough that `n >> d`).
+
+        let mut shuffled: HashSet<u64> = HashSet::with_capacity((n * d) as usize);
+
+        let feistel_keys = &[1, 2, 3, 4];
+        let feistel_precomputed = feistel::precompute((n * d) as feistel::Index);
+
+        for i in 0..n {
+            for k in 0..d {
+                let permuted =
+                    feistel::permute(n * d, i * d + k, feistel_keys, feistel_precomputed);
+                assert!(shuffled.insert(permuted));
```

Comment:
can you add an explanation why this assert has to hold please?

Reply:
indeed i should because that's exactly what i was trying to figure out with these tests

Reply:
done